man/weex.1
.TH weex 1 "2000.05.22" "YN-cxe"


.SH NAME
weex \- fast WEb EXchanger non-interactive FTP client


.SH SYNOPSIS
.B weex
[ Options ]
.I HOSTID [HOSTID...]


.SH DESCRIPTION
.B weex
is a utility designed to automate the task of remotely maintaining a web
page or other FTP archive. With
.B weex
, the maintainer of a web site or archive that must be administered through
FTP interaction can largely ignore that process. The archive administrator
simply creates a local directory that serves as an exact model for the
offsite data. All modifications and direct interaction is done locally to
this directory structure. When the administrator wishes to coordinate the
data on the remote site with that of the local model directory, simply
executing
.B weex
accomplishes this in the most bandwidth-efficient fashion by only
transferring files that need updating. The program will
create or remove (!) files or directories as necessary to accurately
establish the local model on the remote server.
.P
The mandatory
.I HOSTID
argument is the user-defined name that represents a particular FTP account or
configuration specified in the
.I ~/.weexrc
file. This file, the contents of which are described below, can 
alternatively be named
.IR ~/.weex/weexrc .
Multiple
.I HOSTID
arguments to a single
.B weex
command are supported to affect multiple archives or configurations in
immediate succession. In addition, the actual
.I HOSTID
can be substituted with the number representing its relative sequential
position in the
.IR "~/.weexrc" " file (the first " "HOSTID" " definition is 1, and so on)."



.SH OPTIONS
Options at the command line take precedence over any specified in the
configuration file.
.TP
.B -d, --debug-config
Outputs the configuration of each hosts.

.TP
.B -D, --debug-ftplib
Outputs messages from/to FTP server.

.TP
.B -f, --force
If the caching mechanism is inhibiting the uploading of files that should 
be transferred, this option will force the actual transfer to always occur.

.TP
.B -h, --help
Outputs a usage summary to
.IR stdout .

.TP
.B -m, --monochrome
The default mode generates a helpful color coding based on the operation
being performed. This option suppresses that.

.TP
.B -r, --rebuild-cache
If cache file is broken, use this to rebuild it.

.TP
.B -s, --silent
The default mode outputs a helpful status message for each action taken,
including file transfer progress meters.  This mode suppresses output.

.TP
.B -t, --test
Weex doesn't modify any remote files/directories.

.TP
.B -V, --version
Prints the version of the
.B weex
program.

.SH CONFIGURATION FILE
Because this program is meant to be non-interactive, a properly set up
configuration file is essential.
.B weex
looks for this file either in
.I ~/.weexrc
or
.IR ~/.weex/weexrc .
.P
The general format of the configuration file is a bracketed section heading
followed by parameters that define that section's configuration. This
section name is what is provided to
.B weex
at run time to indicate the FTP arrangement you wish to update. A parameter
is only valid for the most recent section heading that preceded it. The
exception to this is the 
.I [default]
section which creates settings for unspecified parameters in every
section. In the case of parameters that accept multiple values, the default
section augments any specifically designated values. 
.P
Lines that begin with the '#' character are safely ignored as comments.
Neither section names nor parameter variables are case sensitive in any
way. Section names can not begin with numbers. Some parameters can be
assigned multiple values. Where this is not possible, the last value is
used. To quote special characters, use single quotes (') or double quotes
("). To quote a particular quote symbol, use the other quote. The general
syntax format for the configuration file is:
.TP 
[hostid1]
 Parameter1 = Value
 # First method for multiple values
 Parameter2 = Value
 Parameter2 = Value
 ...
.TP
[hostid2]
 Parameter1 = Value
 # Second method for multiple values
 Parameter2 = {
        Value
        Value
 }
 ...
.TP
[default]
 Parameter = Value
 ...

.P
.B PARAMETER LIST
.P
These parameters are used to define the properties of a particular host
configuration named in brackets. When specifying directory names, the
trailing "/" is optional. Also, both files and directories (except SrcDir
and DestDir) can be specified with shell wildcards.
When specifying files, if a file or file pattern is given without a path,
it is assumed to be available when encountered in any directory.
If it is an explicitly specified path/file combination, the configuration
is available in that directory.
Files/directories except `SrcDir' and `DestDir' can be specified as
both absolute path and relative path. But you must specify `SrcDir'
and `DestDir' as absolute path.
When you want to specify `/public_html/ignoreme.html' as relative path,
you should specify `./ignoreme.html' rather than `ignoreme.html'.
When specifying binary states, you may use any of the following:
True/False, Yes/No, T/F, 1/0. None are case sensitive.

.TP
.B AsciiFile
Filename patterns that will be transferred using ASCII mode. The default
mode for files not specified in this way is IMAGE(BINARY) mode.

.TP
.B AuthorizationName
If you are using an FTP proxy server that requires challenge/response
authorization with the nonstandard AUTHORIZE and RESPONSE commands,
set this parameter to your authorization user name (or number).  After
logging in to the proxy, weex will prompt you with the challenge and
read back a response.

.TP
.B ChangePermission
The access permissions of the files in the directories specified with
ChangePermissionDir are changed to this parameter. The format is a three
digit octal number.
 
.TP
.B ChangePermissionDir
Files in directories assigned to this parameter will have their
access permissions changed after sending.
Specify local directories.

.TP
.B ChdirAtConnection
Set `true' to change remote current working directory
to DestDir at connection.
If remote current working directory at connection is
DestDir, weex runs faster a bit by setting this parameter
to `false'.
Default is `true'.

.TP
.B ConvToLower
If set to `true', the filenames are converted to lower case before sending.
This feature allows interaction with FTP servers that do not support case
sensitive filenames. If a naming conflict arises due to a lower case name
already existing, an error occurs. Furthermore, the cache is totally
unaware of original uppercase filenames.
 
.TP
.B DestDir
Destination directory on the remote FTP server where the mirrored
information will be sent. This setting corresponds to the "pwd" command on
traditional FTP clients.

.TP
.B Force
If the caching mechanism is inhibiting the uploading of files that should 
be transferred, this option will force the actual transfer to always occur
if set to `true'.

.TP
.B FollowSymlinks
Set `true' to follow symbolic links.
Default is `false'. Symbolic link of directory assumes
a file whose size is 0 byte at default.

.TP
.B FtpPassive
When it is `false', weex uses port mode instead of passive mode.
If the FTP server doesn't support passive mode, specify `false'.
(It may make transfer slower)
Default is `true'.
(See also FAQ)

.TP
.B HostName
Hostname to connect to. The "ftp://" protocol specifier is implied and
shouldn't be used. An IP address is also valid and may save lookup time.

.TP
.B IgnoreLocalDir
This parameter contains one or more protected local directories that are
completely ignored during the transfer process. This is useful if you want
to have subdirectories that are part of a project but don't need to be in
the final, online version (i.e., old versions).

.TP
.B IgnoreLocalFile
This parameter contains one or more protected local file specifiers that
will be completely ignored during the transfer process.

.TP
.B IgnoreRemoteDir
This parameter contains one or more protected remote directories that are
completely ignored during the transfer process. This is useful if you want
to have other material on the remote FTP site that is not related to the
specific project that
.B weex
is dealing with. This could include, for example, data administered through
a different FTP client or process.

.TP
.B IgnoreRemoteFile
This parameter contains one or more protected remote file specifiers that
will be completely ignored during the transfer process.

.TP
.B KeepRemoteDir 
In their directories, files are not removed when they don't exist in the
local directory.

.TP
.B LogDetailLevel
It specifis how detail level weex records a log at.
Connection and disconnection message and error messages are always recorded.
When weex finishes working correctly, it writes `Complete' previous
disconnection message.
If 1 or less, records each connection.
If 2, records each directory.
If 3 or more, records each file manipulated
Default is 1.

.TP
.B LoginName
Login name of the account on the FTP server.

.TP
.B MaxRetryToSend
When weex fails in sending a file, it retry to send the file.
The maximum times of retrying is this parameter.
Default is 8. Specify -1 to stop retrying.

.TP
.B Monochrome
The default mode generates a helpful color coding based on the operation
being performed. A `true' setting suppresses that.

.TP
.B NestSpaces
An integer that controls the indent spacing of the output. The default is 4.
If you want to suppress it, specify -1. If you specify 0, weex uses 4. :-)

.TP
.B OverwriteOK
If the FTP server has trouble overwriting files, set this parameter to `false',
and each file will be deleted before a new version is put in place.

.TP
.B Password
Plaintext password granting access to the account. Note that this file uses
no special facilities to safeguard passwords. Make sure that the file
permissions of a configuration file containing passwords are set
conservatively. If they are not, a warning will be issued at execution.
If this parameter is not found, weex asks for it on the terminal.

.TP
.B PreservePermissionDir
The access permissions of the files in directories
assigned to this parameter will be copied to remote.
Specify *local* directories.

.TP
.B RecordLog
When `true', weex create a log.

.TP
.B RenameOK
If this parameter is set to `true', each file is uploaded under the 
temporary name `weex.tmp', then renamed to its correct name only if
the upload succeeds.  This avoids the problem of a user getting an
incomplete file if he downloads from your Web site while you are
uploading to it, and of incomplete files being uploaded if your FTP
connection breaks while weex is running .  The default is `false'
since some FTP servers might not support renaming.

.TP
.B ShowHiddenFile
Set `true' to get hidden files (starting with a dot) on
FTP server which doesn't show hidden files by default.
Default is `false'.

.TP
.B Silent
The default mode outputs a helpful status message for each action taken,
including file transfer progress meters. A `true' setting suppresses output.

.TP
.B SrcDir
Top of the source directory tree containing the local copy of the
information to be mirrored. This setting corresponds to the "lcd" command
on traditional FTP clients.


.SH EXAMPLES
Here is an example of what a typical
.B .weexrc
file might look like: 

 #-=-=-=-=-=Sample Configuration file=-=-=-=-=- 
 # My favorite FTP account...
 [Ninja]
 HostName = ftp.ninja-rights.org
 LoginName = chrisxed
 Password = '"mYsEcReT!"'
 SrcDir = /home/chrisxed/project/weex
 DestDir = /
 ASCIIfile = *.c
 IgnoreLocalDir = /home/chrisxed/project/weex/weex.devel
 IgnoreLocalDir = /home/chrisxed/project/weex/weex.old
 IgnoreLocalFile = notes2myself.txt
 IgnoreRemoteDir = /offsitearchive/

 # Another FTP account...
 [Veggie]
 HostName = ftp.vegetable-advocacy.com
 LoginNAME = waxedbean
 Password = "X'sBean"
 SrcDir = /home/chrisxed/project/legumes
 DestDir = /souppot/

 # Global configuration settings
 [default]
 AsciiFile = {
 *.htm
 *.html
 *.txt
 *.asc
 }
 IgnoreLocalFile = {
 *.bak
 *.tmp
 *.swp
 }

With a configuration like this, executing:

.B $ weex Ninja

would cause the file system assigned to 
.BR SrcDir " to be completely mirrored on the remote server."
Executing:

.B $ weex ninja veggie

would cause the FTP update to occur for the "Ninja" configuration and
then for the "Veggie" configuration. This could also be specified like
this:

.B $ weex 1 2


.SH NOTES
.P
.B Timestamp Cache Facility
.P
.B weex
implements a cache of file timestamps from Ver 1.6.0.
This makes updating much faster when you have a lot of files.
In addition,
.B weex
also caches directory structure and file names from Ver 2.5.0.
So never slow transfer even if you have many directories and files.
If, however, you change the remote files with another FTP client
or running
.B weex
is killed or
.B weex
is terminated by an error, the cache file will not be correct.
The simple solution to this problem is to remove the cache file (located in
.I ~/.weex/weex.cache.HOSTNAME
) completely or to run
.B weex
with option
.IR --rebuild-cache .


.P
.B Tested Operating Systems
.P
  Debian GNU/Linux 2.1
  Red Hat Linux 5.1/5.2/6.0/6.1J
  SuSE Linux 6.1/6.2/6.3
  Linux Mandrake 6.1
  FreeBSD(98) 3.0-RELEASE
  FreeBSD 3.3-RELEASE
  Solaris 2.5.1


.SH AUTHOR
This very handy program was written by:
.RS
Yuuki NINOMIYA
.RI "<gm@debian.or.jp>"
.RE

The original man page was written by:
.RS
.RI Chris " X " Edwards
.RI "<chrisxed@usa.net>"
.RE

The copyright of this software and documentation belongs to Yuuki NINOMIYA.
It is released under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2,
or (at your option) any later version.
You can redistribute it and/or modify it under the GPL.

This software uses shhopt for parsing command line options.
Shhopt is released under the Artistic License. You may use
it separately under the Artistic License.



doc/ChangeLog
* History

 + 2000/06/13 Ver 2.6.1
   - Support for MACOS type remote server.
   - Bugfix for reporting the wrong number of sent/delected files
     in log file.
   - French translations (fr.po) has been added.
     Thanks to Olivier Tharan <olive@minet.net>.

 + 2000/05/22 Ver 2.6.0
   - Bugfix for trying to remove current directory
     when remove a directory.
   - The way of check of DestDir is changed.
     `It exists if "mkdir <DestDir>" fails.' to
     `It exists if "chdir <DestDir>" successes.'
   - Run "chdir <DestDir>" at connection for safety reasons.
   - A new parameter `ChdirAtConnection' has been added. It
     controls the above.
   - Dutch translations (nl.po) has been added.
     Thanks to Jos Smeets <jsmeets@knoware.nl>.

 + 2000/05/03 Ver 2.5.4
   - A new parameter `FollowSymlinks' has been added. 
     By default, weex doesn't expand symbolic links on local
     for safety reasons. If you need expansion of symbolic links,
     please set this parameter to `true'.
     Thanks to Jeff Abrahamson <abrahj01@molbio.sbphrd.com>.

 + 2000/04/16 Ver 2.5.3
   - Bugfix for parsing remote directory names incorrectly
     with some FTP server based on Windows NT.
     Thanks to Craig Wilkie.

 + 2000/03/24 Ver 2.5.2
   - Even if suppress messages, weex outputs one new line when
     disconnect.
   - Add Spanish translations. (es.po)
     Thanks to Gorka Olaizola <olsago@jet.es>.

 + 2000/02/29 Ver 2.5.1
   - If a password can not be found in the configuration file,
     weex asks for it on the terminal.
     Thanks to Mohammed Waleed Kadous <waleed@cse.unsw.EDU.AU>.

 + 2000/02/25 Ver 2.5.0
   - Directory structure and file names are saved in cache.
     It suppresses LIST command. So it makes transfer much faster.
     WARNING! A cache file of this version is NOT compatible
     with previous version. Please remove all cache files as the
     following before using.

       % rm ~/.weex/weex.cache.*

   - Suppressing CWD command by using absolute path as argument
     of FTP command. It makes transfer much faster.
   - Parameter `SupportFor@Nifty' is removed. Please modify
     a configuration file.
     It isn't needed from this version.
   - Weex outputs `Processing' message when specify multiple hosts
     as command line argument.
   - If there isn't the directory specified as DestDir, create it.
   - Add a new parameter `ShowHiddenFile'. This is for FTP server
     which doesn't show hidden files (starting with a dot) by default.
   - Add a new parameter `PreservePermissionDir'. In these directories,
     weex copies access permission of the local file to the remote one.
   - Add a new option `--rebuild-cache'. If cache file is broken, use this.
   - Add a new option `--debug-ftplib'. Output messages from/to FTP server.
   - Update de.po.
     Thanks to Christian Ullrich <chris@chrullrich.de>.

 + 2000/01/09 Ver 2.3.0
   - The bug is fixed that weex crashes when there is a parameter
     before defining section in a configuration file.
   - Add a new option `--test' that doesn't actually change
     remote files/directories.
   - The bug is fixed that a cache file is broken when weex processes
     the file which included blank in its name.
   - When weex changes the current directory, it removes
     the last slash of directory name.
   - Support for the FTP server which doesn't output group name when `ls'.
   - Support for port mode for the FTP server which doesn't support
     passive mode by a new parameter `FtpPassive'. 
   - Update de.po.
     Thanks to Christian Ullrich <chris@chrullrich.de>.

 + 1999/12/20 Ver 2.2.0
   - Support for configuration of files/directories (except `SrcDir'
     and `DestDir') as relative path.
   - Add a new parameter `KeepRemoteDir'.
     Files in this directory are not removed when they don't exist
     in the local directory.
   - Support for Geocities. It is recognized automatically.

 + 1999/12/11 Ver 2.1.2
   - The bug is fixed that processing a symlink is wrong again.
     Thanks to Christian Ullrich <chris@chrullrich.de>.

 + 1999/12/10 Ver 2.1.1
   - Update de.po.
     Thanks to Christian Ullrich <chris@chrullrich.de>.
   - Some messages are modified.
   - The bug is fixed that processing a symlink is wrong.

 + 1999/12/09 Ver 2.1.0
   - Update de.po.
     Thanks to Christian Ullrich <chris@chrullrich.de>.
   - Support for Solaris.
   - Support for a file whose name includes spaces.
   - The parameter `LsAtNifty' is changed to `SupportFor@Nifty'.
     Please type as the following.

       % cd ~/.weex
       % mv weexrc weexrc.old
       % sed -e "s/LsAtNifty/SupportFor@Nifty/" < weexrc.old > weexrc

   - Weex continues processing if it receives a minor error.
   - Retry to send when weex fails in sending a file.
     The maximum times of retrying can be specified by `MaxRetryToSend'.
   - Add a logging facility. It would be useful for using weex with cron
     or background execution.

 + 1999/11/29 Ver 2.0.7
   - Support for Windows_NT type remote server.
   - The file name of the temporary file stored file lists on remote
     server is made by tempnam().
     You can use multiple weex at the same time.

 + 1999/11/25 Ver 2.0.6
   - Modify documents.
   - Rewrite set_default() in config.c.
   - The following items are for supporting FreeBSD.
     Thanks to Julian C. Dunn <jdunn@aquezada.com>.
     Uses gnuregex.h prior to regex.h.
   - If constant macro `PATH_MAX' is undefined, define as 512.
   - Uses strtok() instead of sscanf() because a compiler on FreeBSD
     outputs warnings when use `%as' in sscanf().
   - The bug is fixed that the progress information is always 0.

 + 1999/11/14 Ver 2.0.5a
   - Update de.po.
     Thanks to Christian Ullrich <chris@chrullrich.de>.

 + 1999/11/13 Ver 2.0.5
   - The bug is fixed that the file at the first of directory entry
     on remote side is ignored. It caused that the file is not sent
     or weex outputs `Cannot make directory `foo'.' and exit.

src/weex.c
/**************************************************************************/
/*                                                                        */
/*  weex - a non-interactive mirroring utility for updating web pages     */
/*  Copyright (C) 1999-2000 Yuuki NINOMIYA <gm@debian.or.jp>              */
/*                                                                        */
/*  This program is free software; you can redistribute it and/or modify  */
/*  it under the terms of the GNU General Public License as published by  */
/*  the Free Software Foundation; either version 2, or (at your option)   */
/*  any later version.                                                    */
/*                                                                        */
/*  This program is distributed in the hope that it will be useful,       */
/*  but WITHOUT ANY WARRANTY; without even the implied warranty of        */
/*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         */
/*  GNU General Public License for more details.                          */
/*                                                                        */
/*  You should have received a copy of the GNU General Public License     */
/*  along with this program; if not, write to the                         */
/*  Free Software Foundation, Inc., 59 Temple Place - Suite 330,          */
/*  Boston, MA 02111-1307, USA.                                           */
/*                                                                        */
/**************************************************************************/

/* $Id: weex.c,v 1.2 2004/03/22 05:57:19 gmkun Exp $ */

#define GLOBAL_VALUE_DEFINE

#ifdef HAVE_CONFIG_H
#  include "config.h"
#endif

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/stat.h>
#include <sys/types.h>

#ifdef ENABLE_NLS
#  include <locale.h>
#endif

#include "intl.h"
#include "ftplib.h"
#include "shhopt.h"
#include "strlib.h"
#include "variable.h"
#include "proto.h"


static void put_usage_and_exit(void);
static void put_version_and_exit(void);
static void make_weex_directory(void);


static optStruct option_table[] = {
	/* short long               type        var/func                          special */
	{ 'a', "all-hosts",         OPT_FLAG,   &command_line_option.all_hosts,         0 },
	{ 'c', "config-file",       OPT_STRING, &command_line_option.config_file,       0 },
	{ 'q', "quiet",             OPT_FLAG,   &command_line_option.quiet,             0 },
	{ 'p', "suppress-progress", OPT_FLAG,   &command_line_option.suppress_progress, 0 },
	{ 'f', "force",             OPT_FLAG,   &command_line_option.force,             0 },
	{ 'm', "monochrome",        OPT_FLAG,   &command_line_option.monochrome,        0 },
	{ 'r', "rebuild-cache",     OPT_FLAG,   &command_line_option.rebuild_cache,     0 },
	{ 'R', "catch-up",          OPT_FLAG,   &command_line_option.catch_up,          0 },
	{ 'w', "not-mkdir-weex",    OPT_FLAG,   &command_line_option.not_mkdir_weex,    0 },
	{ 'C', "confirmation",      OPT_FLAG,   &command_line_option.confirmation,      0 },
	{ 's', "simulate",          OPT_FLAG,   &command_line_option.simulate,          0 },
	{ 'l', "list",              OPT_FLAG,   &command_line_option.list,              0 },
	{ 'n', "nlist",             OPT_FLAG,   &command_line_option.nlist,             0 },
	{ 'v', "view",              OPT_FLAG,   &command_line_option.view,              0 },
	{ 'd', "download",          OPT_FLAG,   &command_line_option.download,          0 },
	{ 'u', "upload",            OPT_FLAG,   &command_line_option.upload,            0 },
	{ 'g', "debug",             OPT_FLAG,   &command_line_option.debug,             0 },
	{ 'h', "help",              OPT_FLAG,   put_usage_and_exit,          OPT_CALLFUNC },
	{ 'V', "version",           OPT_FLAG,   put_version_and_exit,        OPT_CALLFUNC },
	{ 0, 0, OPT_END, 0, 0 }
};


/* --- MAIN FUNCTION --- */

int main(int argc, char *argv[])
{
	int max_hosts;

#ifdef ENABLE_NLS
	setlocale(LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);
#endif

	optParseOptions(&argc, argv, option_table, 0);

	if (!command_line_option.not_mkdir_weex) {
		make_weex_directory();
	}

	if (command_line_option.debug) {
		ftplib_debug = 3;
	}

	if (command_line_option.config_file == NULL) {
		command_line_option.config_file = str_concat(getenv("HOME"), "/", DEFAULT_CONFIG_DIR, "/", DEFAULT_CONFIG_FILE, NULL);
	}

	max_hosts = parse_config_file();

	if (command_line_option.view) {
		display_host_definition(argc, argv, max_hosts);
		exit(0);
	}

	if (argc < 2 && !command_line_option.all_hosts) {	/* No host is specified. */
		put_usage_and_exit();
	}

	main_loop(argc, argv, max_hosts);

	return (0);
}


/* --- PUBLIC FUNCTIONS --- */

void internal_error(const char *file, int line)
{
	fprintf(stderr, _("Internal error has occurred in %s at line %d.\n"), file, line);
	fprintf(stderr, _("It must be a bug. Please send a bug report to the author.\n"));
	exit(1);
}


void minor_error(int host)
{
	if (config.exit_on_minor_error[host]) {
		disconnect_from_remote_host(host);
		exit(1);
	}
}


void fatal_error(int host)
{
	disconnect_from_remote_host(host);
	exit(1);
}


/* --- PRIVATE FUNCTIONS --- */

static void put_usage_and_exit(void)
{
	printf(_("weex version %s Copyright (C) 1999-2001 Yuuki NINOMIYA <gm@debian.or.jp>\n\n"), VERSION);
	printf(_("This program comes with ABSOLUTELY NO WARRANTY, to the extent permitted by law.\n"
		 "You may redistribute it under the terms of the GNU General Public License.\n"
		 "For more information about these matters, see the file named COPYING.\n\n"));
	printf(_("usage: weex [OPTIONS] HOST...\n\n"));
	printf(_("options:\n"));
	printf(_("  -c, --config-file=FILE   use an alternate configuration file\n"));
	printf(_("  -w, --not-mkdir-weex     not make ~/.weex directory on startup\n"));
	printf(_("  -v, --view               display the host difinitions\n"));
	printf(_("  -l, --list               display a listing of files needing update\n"));
	printf(_("  -n, --nlist              display the number of files needing update\n"));
	printf(_("  -r, --rebuild-cache      rebuild a cache file\n"));
	printf(_("  -R, --catch-up           rebuild a cache file as all files are updated\n"));
	printf(_("  -u, --upload             transfer files from local to remote\n"));
	printf(_("  -d, --download           transfer files from remote to local\n"));
	printf(_("  -a, --all-hosts          process all defined hosts\n"));
//	printf(_("  -C, --confirmation       request confirmation when change files\n"));
	printf(_("  -f, --force              transfer all files whether updated or not\n"));
//	printf(_("  -s, --simulate           perform ordering simulation (not actually change)\n"));
	printf(_("  -q, --quiet              suppress non-error messages\n"));
	printf(_("  -p, --suppress-progress  suppress progress information\n"));
	printf(_("  -m, --monochrome         output messages in monochromee\n"));
	printf(_("  -g, --debug              enable debug output\n"));
	printf(_("  -h, --help               display this help and exit\n"));
	printf(_("  -V, --version            display version information and exit\n"));
	printf("\n");

	exit(0);
}


static void put_version_and_exit(void)
{
	printf(_("weex %s\nCopyright (C) 1999-2001 Yuuki NINOMIYA <gm@debian.or.jp>\n\n"), VERSION);
	printf(_("This program comes with ABSOLUTELY NO WARRANTY, to the extent permitted by law.\n"
		 "You may redistribute it under the terms of the GNU General Public License.\n"
		 "For more information about these matters, see the file named COPYING.\n\n"));

	exit(0);
}


static void make_weex_directory(void)
{
	char *temp;

	temp = str_concat(getenv("HOME"), "/.weex", NULL);
	if (access(temp, X_OK) != 0) {
		if (mkdir(temp, 0755) != 0) {
			perror("mkdir");
			fprintf(stderr, _("Cannot make directory `%s'.\n"), temp);
			exit(1);
		}
		fprintf(stderr, _("Directory `%s' has been made.\n"), temp);
	}
	free(temp);
}
